### DISCUSSION TITLE
Add delete all row function!!

### DISCUSSION THREAD

**Discussion by wpdbs1229** [Akascape/CTkTable#113] [Ideas]
**Hi, friends** 😃 

I thought it would be convenient to have a function to remove all rows except the header when using CTk Table, so I made one.

You can use it by adding the code below to ctktable.py!
```
def get_row_count(self):
        return len(self.values)
    
# <---- Doesn't include headers ---->
def delete_all_row(self):
        end_row_idx  = self.get_row_count()
        self.delete_rows(list(range(1, end_row_idx)))
        
```

Here's an **example code**  👍 
```
import customtkinter
from CTkTable import *

root = customtkinter.CTk()


def delete_all_data():
    table.delete_all_row()
        
value = [["A","B","C","D","E"],[1,2,3,4,"▇▇"],[0,0,0,0,"▇▇"],[1,2,0,0,"▇▇"],[1,2,3,4,"▇▇"]]

table = CTkTable(master=root, row=5, column=5, values=value, command=show ,header_color="green")
table.pack(expand=True, fill="both", padx=20, pady=20)

    
btn = customtkinter.CTkButton(root, text='delete all data', width=150, command=delete_all_data)
btn.pack(expand=True,pady=5, padx=5)
root.mainloop()
```

If you have any opinions, feel free to tell me!!! ✍️

**Comment by OLIVERFRANCISCO** [comment]
Te falto el metodo de show. . . 
Aunque tambien lo que se podria hacer si quieres meter acciones a las registros, etc
es crear un frame para cada registro donde se podria meter botones con funciones lambda 
y ya , 
la tabla estaria contenida en la raiz o el CTk() o root
un ejemplo de lo que te quiero decir. . . 

      `import customtkinter as ctk
      
      class TablaCortes:
          def __init__(self, master):
              self.master = master
              self.frame_inferior = ctk.CTkFrame(master)
              self.frame_inferior.pack(expand=True, fill="both", pady=(10, 10), padx=(10, 10))
              self.tabla_cortes()
      
          def tabla_cortes(self):
              self.values = []
              self.values.append(["Animal", "Nombre del corte", "Acciones"])
      
              try:
                  # Simulación de datos
                  self.cortes = [
                      (1, "Cerdo", "Lomo", "Lomo de cerdo"),
                      (2, "Cerdo", "Costilla", "Costilla de cerdo"),
                      (3, "Cerdo", "Panceta", "Panceta de cerdo")
                  ]
      
                  for corte in self.cortes:
                      # Crear un frame para cada fila
                      row_frame = ctk.CTkFrame(self.frame_inferior)
                      row_frame.pack(fill="x", pady=(0, 5))
      
                      # Añadir los valores de las columnas
                      ctk.CTkLabel(row_frame, text=corte[1]).pack(side="left", padx=(10, 0))
                      ctk.CTkLabel(row_frame, text=corte[3]).pack(side="left", padx=(10, 0))
      
                      # Añadir el botón de acciones
                      ctk.CTkButton(row_frame, text="Prueba", command=lambda c=corte: print(f"Hola {c[3]}")).pack(side="right", padx=(10, 0))
      
              except Exception as e:
                  print(e)
      
      if __name__ == "__main__":
          root = ctk.CTk()
          app = TablaCortes(root)
          root.mainloop()`

**Reply by wpdbs1229** [reply]
Oh, thank you so much for the feedback! 😊

I designed delete_all_rows() to safely remove all data rows except the header.
Each row is managed as a Frame, so when deleting, I destroy the Frame and reindex the remaining rows to maintain the structure.
After deletion, refresh_table() can be used to reset the layout if needed.

Managing rows as Frames makes it easier to extend the table later (like adding action buttons, inline editing, etc.).
```
def delete_all_rows(self):
    # Delete all data rows from the table (excluding the header).
    row_count = self.get_row_count()

    if row_count <= 1:
        return  # No data rows to delete

    # Create a list of row indices to delete (starting from index 1)
    row_indices = list(range(1, row_count))

    for row_index in row_indices:
        if row_index in self.rows:
            row_frame = self.rows[row_index]
            row_frame.destroy()    # Destroy each row (Frame)
            del self.rows[row_index]

    self.reindex_rows()
    self.refresh_table()

def reindex_rows(self):
    # Reindex the remaining rows after deletion.
    new_rows = {}
    for new_index, frame in enumerate(self.rows.values(), start=1):
        new_rows[new_index] = frame
        frame.grid(row=new_index, column=0, sticky="nsew")  # Reposition frames if necessary

    self.rows = new_rows
```
Here's a simple example of how you can connect delete_all_rows() to a button
Example: Create a "Delete All" button that clears all rows
```
import customtkinter as ctk

class App(ctk.CTk):
    def __init__(self):
        super().__init__()

        self.title("CTkTable Example")
        self.geometry("600x400")

        # Initialize the table (assuming YourCTkTable has delete_all_rows method)
        self.table = YourCTkTable(self)
        self.table.pack(pady=20)

        # Create a "Delete All Rows" button
        delete_button = ctk.CTkButton(self, text="Delete All Rows", command=self.table.delete_all_rows)
        delete_button.pack(pady=10)

if __name__ == "__main__":
    app = App()
    app.mainloop()
```
